conan/tools/microsoft/msbuild.py

```diff
@@ -27,10 +32,20 @@ def __init__(self, conanfile):
         if conanfile.settings.get_safe("os") == "WindowsCE":
             msvc_arch = conanfile.settings.get_safe("os.platform")
         self.platform = msvc_arch
+        self.toolset = msvs_toolset(conanfile)
 
     def command(self, sln, targets=None):
-        cmd = ('msbuild "%s" /p:Configuration=%s /p:Platform=%s'
-               % (sln, self.build_type, self.platform))
+        cmd = (f'msbuild "{sln}" /p:Configuration={self.build_type} '
+               f"/p:Platform={self.platform} /p:PlatformToolset={self.toolset}")
```

Comment:
This adds back the PlatformToolset as argument only, and that doesn't work for user consumption, that will only be loading the toolchain. That configuration needs to be in the toolchain file.

Reply:
I've not removed PlatformToolset from toolchain. It's added as an argument to msbuild because when `conantoolchain.props` is injected with `/p:ForceImportBeforeCppTargets`, I've seen that PlatformToolset from this toolchain was ignored.

Reply:
Actually, from my experimentation when `conantoolchain.props` is injected with `/p:ForceImportBeforeCppTargets`:
- Everything under `ItemDefinitionGroup` is honored.
- Everything under `PropertyGroup` (`PlatformToolset` & custom properties) is ignored by MSBuild.

This is why I inject `p:PlatformToolset` in msbuild command line.

Reply:
https://learn.microsoft.com/en-us/cpp/build/project-property-inheritance

> When you set a property with property pages, it's also important to pay attention to where you set it. If you set a property to "X" in a .props file, but the property is set to "Y" in the project file, then the project will build with the property set to "Y".

Reply:
The problem with ``/p:ForceImportBeforeCppTargets`` is the developer experience. Most developers with VS will just double click the project or solution Whatever is being passed should be in the ``vcxproj`` by default, it cannot rely on cli args to work. This is probably the most massive use case of this integration, there are companies out there using it, and hundreds of developers do this daily. So this cannot be changed if this doesn't keep working as-is (this is why I suggest an opt-in for ConanCenter recipes, which are typically not opened by developers to touch the source code)

Reply:
There is no breaking change, I've not changed toolchain files.

Here is another proposal:

- add something like `props_autoinjection` argument to `build()` method of `MSBuild`, `False` by default.
- add properties dict attribute to `MSBuild`
- when `props_autoinjection` is False, current behavior is kept, so no breaking change at all.
- when `props_autoinjection` is True:
  - it collects root props of `MSBuildToolchain` & `MSBuildDeps` and inject them to `/p:ForceImportBeforeCppTargets`.
  - it also injects `/p:<property>=<value>` to msbuild command line
- keep previous tests as is, an write other for the `props_autoinjection=True` flavor.

Reply:
Regarding properties, to avoid duplication between toolchain and MSBuild, there is a more complex solution: gather all properties by parsing all props programmatically, and inject them in command line if props_autoinjection is True.

Reply:
Yes, sorry, I meant the ``/p:ForceImportBeforeCppTargets`` is a breaking change when applied to ``conandeps.props``. For the toolchain, the main issue is that it can easily be hiding errors for developers opening the IDE

Yes, I think this approach would be better. Just a couple of suggestions:
- Make explicit the intention, ``props_autoinjection`` is a bit generic, something like ``force_imports`` or ``force_import_generated_files``
- Generic property injection is a different feature and ``MSBuildToolchain`` already provides injection of properties via toolchain (so they work for developers opening the IDE, without needing the Conan ``build()`` injection), so this should be a separate discussion if there is a use case for it.

Reply:
> Generic property injection is a different feature and MSBuildToolchain already provides injection of properties via toolchain (so they work for developers opening the IDE, without needing the Conan build() injection), so this should be a separate discussion if there is a use case for it.

As I said in https://github.com/conan-io/conan/pull/12817#discussion_r1059935272, properties defined in toolchain files are not honored when injected through `/p:ForceImportBeforeCppTargets`, so for me it's related. `/p:ForceImportBeforeCppTargets` would be an internal detail of `force_import_generated_files =True`, the goal of this argument is to honor everything defined in these props, by any means.